php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+     */
+    public function handleSignal(string $input): string 
+    {
+        // ...
+    }
+}
+``` 
+
+> The actual annotation format is TBD as it irrelevant to the implementation and might vary from framework to framework. 
+
+### Syntax and Atomic Blocks
+The building blocks of workflow is `yield` commands which might return actual value or a promise for this value. Every
+command wrapped via dedicated class:
+
+```php
+function run()
+{
+    yield new Activity('name', 'values', [/* options */]);
+}
+```
+
+Some commands might automatically return sub-commands. For example blocking activity execution:
+
+```php
+$result = yield (new Activity('name', 'values', [/* options */]))->get();
+```
+
+To wait for multiple activities at the same time we can introduce higher level promises such as:
+
+```php
+$a = yield new Activity('a', 'values', [/* options */]);
+$b = yield new Activity('b', 'values', [/* options */]);
+
+yield new WaitAll($a, $b);
+
+// or 
+yield new WaitAny($a, $b);
+```
+
+> More atomic blocks can be added down the road.
+
+#### Syntax Sugar
+It should be possible to mock `new Activity` and other commands with shorted methods from abstract workflow implementation:
+
+```php
+class MyWorkflow extends Workflow 
+{
+    public function run()
+    {
+        $result = yield $this->activities->doSomething()->get();
+        yield $this->sleep(1);
+
+        // ...
+    }
+}
+``` 
+
+Examples below will be given using sugared approach for simplicity (it is possible to fallback to commands as well). 
+
+#### Timers
+Workflow timers can operate as promise or blocking operation:
+
+```php
+$timer = yield $this->timer(1 * Time::DAY);
+
+yield $timer->wait();
+```
+
+Blocking:
+
+```php
+yield $this->timer(1 * Time::DAY)->wait();
+
+// or
+yield $this->sleep(1 * Time::DAY);
+```
+
+Can be comfined with Promises:
+
+```php
+$a = yield $this->activities->doSomething(...);
+$t = yield $this->timer(1 * Time::MINUTE);
+
+yield $this->waitAny($a, $t);
+
+if ($t->isFired()) {
+    // ...
+}
+```
+
+### Queries
+Workflow queries can be easily implemented using dedicated query method of workflow. The state of the workflow can be 
+extracted from the properties of the object.
+
+```php
+use Temporal\Workflow;
+
+class DemoWorkflow extends Workflow\Workflow
+{
+    private $step = 0;
+
+    public function queryStep(): int 
+    {
+        return $this->step;
+    }
+
+    public function run(string $input): string
+    {
+        yield $this->activities->doSomething($input);
+        $this->step++;
+
+        // ...
+    }
+}
+```
+
+> Query methods must be registered explicitly.
+
+An alternative approach can be based on registering the callback function during the workflow execution:
+
+```php
+yield new QueryHandler('something', function() use (&$step) {
+    return $step;
+});
+```
+
+> Such an approach is more suitable for dynamic workflows. The principle of work is identical to pre-registered queries.
```

Comment:
In Java we support both. Dynamic approach for both query and signals is needed for libraries which don't have access to the top level workflow object.

Reply:
Yep, both should be possible at the same time. Not allowing dynamic workflows in PHP would be a weird decision.